coderd/wspubsub/wspubsub.go

```diff
@@ -0,0 +1,91 @@
+package wspubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+)
+
+// WorkspaceEventChannel can be used to subscribe to events for
+// workspaces owned by the provided user ID.
+func WorkspaceEventChannel(ownerID uuid.UUID) string {
+	return fmt.Sprintf("workspace_owner:%s", ownerID)
+}
+
+func HandleWorkspaceEvent(cb func(ctx context.Context, payload WorkspaceEvent)) func(ctx context.Context, message []byte) {
+	return func(ctx context.Context, message []byte) {
+		var payload WorkspaceEvent
+		if err := json.Unmarshal(message, &payload); err != nil {
+			return
+		}
+		cb(ctx, payload)
+	}
+}
+
+type WorkspaceEvent struct {
+	Kind        WorkspaceEventKind `json:"kind"`
+	WorkspaceID uuid.UUID          `json:"workspace_id" format:"uuid"`
+
+	// WorkspaceName is only set for WorkspaceEventTypeStateChange
+	WorkspaceName *string `json:"workspace_name"`
+	// Transition is only set for WorkspaceEventTypeStateChange
+	Transition *database.WorkspaceTransition `json:"transition,omitempty"`
+	// JobStatus is only set for WorkspaceEventTypeStateChange
+	JobStatus *database.ProvisionerJobStatus `json:"job_status,omitempty"`
```

Comment:
This smells to me like you're going to use these values to determine the state updates that we send over the API. That's a bad idea because PubSub updates are not strictly ordered with respect to database transactions.  So you could have an update setting the JobStatus to Running, then an update setting it to Succeeded, but the PubSub messages could come Succeeded, Running.

The PubSub should not be relied on to give you the latest state, just the kick to go and query the database for the state.